### PR TITLE
React upgrade 16.9

### PR DESCRIPTION
**Pull request by cartogram**
This pull closes https://github.com/Shopify/web/issues/13363 by upgrading to React@16.9.0.alpha0, this includes:

- Upgrading `react` and `react-dom` (note, there are no tags for the updated types yet)
- Modifying relevant parts of the `react-testing` library
- Found all skipped tests waiting on this upgrade and fixed them
- Removed the utility that whitelisted react errors from the console

🎩 There should be no failing tests, no async-act related skipped tests and no act-related console errors.

### CODE REVIEW

File: packages/enzyme-utilities/src/index.ts [hunk 1]
```diff
@@ -8,6 +8,8 @@ export type AnyWrapper =
   | CommonWrapper<any, any>
   | CommonWrapper<any, never>;
 
+const act = reactAct as (func: () => void | Promise<void>) => Promise<void>;
```
Comment (lemonmade) [inline packages/enzyme-utilities/src/index.ts]: Should leave a comment for why this is needed

File: packages/enzyme-utilities/src/index.ts [hunk 2]
```diff
@@ -33,19 +35,23 @@ export function trigger(wrapper: AnyWrapper, keypath: string, ...args: any[]) {
...
-  updateRoot(wrapper);
+    if (isPromise(returnValue)) {
+      return (returnValue as unknown) as Promise<void>;
+    }
+  });
 
-  if (returnValue instanceof Promise) {
-    return returnValue.then(ret => {
+  if (isPromise(returnValue)) {
+    return Promise.resolve(promise as Promise<any>).then(ret => {
```
Comment (lemonmade) [inline packages/enzyme-utilities/src/index.ts]: Should leave a comment of why you check that `returnValue` is a promise, not `promise`

File: packages/enzyme-utilities/src/index.ts [hunk 3]
```diff
@@ -56,3 +62,11 @@ function updateRoot(wrapper: AnyWrapper) {
 export function findById(wrapper: ReactWrapper<any, any>, id: string) {
   return wrapper.find({id}).first();
 }
+
+function isPromise<T>(promise: T | Promise<T>): promise is Promise<T> {
+  return (
+    promise != null &&
+    typeof promise === 'object' &&
+    (promise as any).then != null
```
Comment (lemonmade) [inline packages/enzyme-utilities/src/index.ts]: `'then' in promise`

File: packages/react-form/src/hooks/list/list.ts [hunk 4]
```diff
@@ -216,6 +216,9 @@ export function useList<Item extends object>(
                 }
                 validate();
               },
+              setError(error: string) {
+                dispatch(updateErrorAction({target, error}));
+              },
```
Comment (lemonmade) [inline packages/react-form/src/hooks/list/list.ts]: Why this change?

File: packages/react-form/src/hooks/test/form.test.tsx [hunk 5]
```diff
@@ -156,18 +160,18 @@ describe('useForm', () => {
         .find(TextField, {label: 'title'})!
         .trigger('onChange', 'tortoritos, the chip for turtles!');
 
-      wrapper
-        .find('button', {type: 'submit'})!
-        .trigger('onClick', clickEvent());
+      await wrapper.act(() => {
```
Comment (lemonmade) [inline packages/react-form/src/hooks/test/form.test.tsx]: Not waiting on anything

File: packages/react-form/src/hooks/test/form.test.tsx [hunk 6]
```diff
@@ -177,18 +181,19 @@ describe('useForm', () => {
...
-      wrapper
-        .find('button', {type: 'submit'})!
-        .trigger('onClick', clickEvent());
-
-      await promise;
+      await wrapper.act(async () => {
+        await wrapper
+          .find('button', {type: 'submit'})!
+          .trigger('onClick', clickEvent());
+        await promise;
```
Comment (lemonmade) [inline packages/react-form/src/hooks/test/form.test.tsx]: Do you need to await both of these?

File: packages/react-form/src/hooks/test/form.test.tsx [hunk 7]
```diff
@@ -200,35 +205,38 @@ describe('useForm', () => {
 
       wrapper.find(TextField, {label: 'title'})!.trigger('onChange', '');
 
-      wrapper
-        .find('button', {type: 'submit'})!
-        .trigger('onClick', clickEvent());
+      await wrapper.act(() => {
```
Comment (lemonmade) [inline packages/react-form/src/hooks/test/form.test.tsx]: Not awaiting anything

File: packages/react-form/src/hooks/test/form.test.tsx [hunk 8]
```diff
@@ -200,35 +205,38 @@ describe('useForm', () => {
...
+          .trigger('onClick', clickEvent());
+      });
 
       expect(wrapper).toContainReactComponent('p', {
         children: error.message,
       });
     });
 
-    // Async act() issues block this test https://github.com/facebook/react/issues/15379
+    // This test goes into an infinite render loop
```
Comment (lemonmade) [inline packages/react-form/src/hooks/test/form.test.tsx]: Do you have an issue/ plan for it?

File: packages/react-form/src/hooks/test/form.test.tsx [hunk 9]
```diff
@@ -277,27 +291,30 @@ describe('useForm', () => {
...
         .trigger('onClick', clickEvent());
 
-      await wrapper
-        .find('button', {type: 'reset'})!
-        .trigger('onClick', clickEvent());
+      await wrapper.act(async () => {
+        await wrapper
+          .find('button', {type: 'button'})!
+          .trigger('onClick', clickEvent());
+        await promise;
```
Comment (lemonmade) [inline packages/react-form/src/hooks/test/form.test.tsx]: Do you need both?

File: packages/react-graphql/src/hooks/tests/mutation.test.tsx [hunk 10]
```diff
@@ -63,7 +59,11 @@ describe.skip('useMutation', () => {
     );
 
     mockMutation.find('button')!.trigger('onClick', undefined as any);
-    await graphQL.resolveAll();
+
+    // @ts-ignore
+    await act(async () => {
```
Comment (lemonmade) [inline packages/react-graphql/src/hooks/tests/mutation.test.tsx]: Can we not use `mockMutation.act` here?

File: packages/react-import-remote/src/test/ImportRemote.test.tsx [hunk 11]
```diff
@@ -45,30 +45,50 @@ describe('<ImportRemote />', () => {
...
-    it('uses the props as arguments for load()', () => {
+    it('uses the props as arguments for load()', async () => {
+      const resolve = createResolvableLoad();
       const nonce = '';
-      mount(<ImportRemote {...mockProps} />);
+      const importRemote = mount(<ImportRemote {...mockProps} />);
+
+      await importRemote.act(async () => {
+        await resolve();
+      });
```
Comment (lemonmade) [inline packages/react-import-remote/src/test/ImportRemote.test.tsx]: Why this change?

File: packages/react-import-remote/src/test/ImportRemote.test.tsx [hunk 12]
```diff
@@ -77,27 +97,31 @@ describe('<ImportRemote />', () => {
 
       const newSource = 'https://bar.com/foo.js';
 
-      importRemote.setProps({source: newSource});
+      // we need to have the async here otherwise we will get
+      // a console error an `act(...)` here
+      // eslint-disable-next-line require-await
+      await importRemote.act(async () => {
```
Comment (lemonmade) [inline packages/react-import-remote/src/test/ImportRemote.test.tsx]: It doesn't make sense that this would work

File: packages/react-import-remote/src/test/ImportRemote.test.tsx [hunk 13]
```diff
@@ -77,27 +97,31 @@ describe('<ImportRemote />', () => {
...
 
-      const importRemote = await mount(
+      const importRemote = mount(
         <ImportRemote {...mockProps} onImported={spy} />,
       );
 
-      importRemote.act(async () => {
-        await promise;
+      await importRemote.act(async () => {
+        await resolve;
```
Comment (lemonmade) [inline packages/react-import-remote/src/test/ImportRemote.test.tsx]: Not called as a function here?

Comment (lemonmade) [reply packages/react-import-remote/src/test/ImportRemote.test.tsx]: Makes me think that this is kind of a placebo, and all it's doing is pushing the act to include the next event loop tick?

File: packages/react-import-remote/src/test/ImportRemote.test.tsx [hunk 14]
```diff
@@ -123,37 +147,47 @@ describe('<ImportRemote />', () => {
...
       const importRemote = mount(
         <ImportRemote {...mockProps} defer={DeferTiming.Idle} />,
       );
 
       expect(load).not.toHaveBeenCalled();
 
-      importRemote.act(() => {
-        requestIdleCallback.runIdleCallbacks();
+      await importRemote.act(async () => {
+        await requestIdleCallback.runIdleCallbacks();
```
Comment (lemonmade) [inline packages/react-import-remote/src/test/ImportRemote.test.tsx]: Does this return a promise?

File: packages/react-import-remote/src/test/ImportRemote.test.tsx [hunk 15]
```diff
@@ -164,8 +198,8 @@ describe('<ImportRemote />', () => {
 
       expect(load).not.toHaveBeenCalled();
 
-      importRemote.act(() => {
-        intersectionObserver.simulate({
+      await importRemote.act(async () => {
+        await intersectionObserver.simulate({
```
Comment (lemonmade) [inline packages/react-import-remote/src/test/ImportRemote.test.tsx]: Same question as above

File: packages/react-testing/src/element.ts [hunk 16]
```diff
@@ -173,7 +173,7 @@ export class Element<Props> {
   trigger<K extends FunctionKeys<Props>>(
     prop: K,
     ...args: DeepPartialArguments<Arguments<Props[K]>>
-  ): ReturnType<NonNullable<Props[K]>> {
+  ): Promise<ReturnType<NonNullable<Props[K]>>> {
```
Comment (lemonmade) [inline packages/react-testing/src/element.ts]: Please revert, this is not correct

File: packages/react-testing/src/element.ts [hunk 17]
```diff
@@ -187,7 +187,7 @@ export class Element<Props> {
     });
   }
 
-  triggerKeypath<T = unknown>(keypath: string, ...args: unknown[]): T {
+  triggerKeypath<T = unknown>(keypath: string, ...args: unknown[]): Promise<T> {
```
Comment (lemonmade) [inline packages/react-testing/src/element.ts]: Same as above

File: packages/react-testing/src/root.tsx [hunk 18]
```diff
@@ -76,26 +77,34 @@ export class Root<Props> {
...
+      result = action();
+
+      if (isPromise(result)) {
+        return (result as unknown) as Promise<void>;
+      }
+    });
 
-    return result;
+    if (isPromise(result)) {
+      return Promise.resolve(promise).then(afterResolve) as any;
```
Comment (lemonmade) [inline packages/react-testing/src/root.tsx]: Same request as above re: clarifying comments for this and the type casting of `act` above

File: packages/react-testing/src/root.tsx [hunk 19]
```diff
@@ -287,3 +296,11 @@ function childrenToTree(fiber: Fiber | null, root: Root<unknown>) {
 
   return {children, descendants};
 }
+
+function isPromise<T>(promise: T | Promise<T>): promise is Promise<T> {
+  return (
+    promise != null &&
+    typeof promise === 'object' &&
+    (promise as any).then != null
```
Comment (lemonmade) [inline packages/react-testing/src/root.tsx]: `'then' in promise`

File: packages/react-testing/src/root.tsx [hunk 20]
```diff
@@ -141,7 +150,7 @@ export class Root<Props> {
   trigger<K extends FunctionKeys<Props>>(
     prop: K,
     ...args: DeepPartialArguments<Arguments<Props[K]>>
-  ): ReturnType<NonNullable<Props[K]>> {
+  ): Promise<ReturnType<NonNullable<Props[K]>>> {
```
Comment (lemonmade) [inline packages/react-testing/src/root.tsx]: Please remove

File: packages/react-form/src/hooks/list/list.ts (hunk 4 above)
Comment (TheMallen) [reply packages/react-form/src/hooks/list/list.ts]: The test for submitting list failed because the imperative method for setting an error on list fields was missing (probably didn't pull it out of my codesandbox version).

Comment (cartogram) [reply packages/react-form/src/hooks/list/list.ts]: This implementation was missing but it was only caught when I un-skipped the form tests. I can open a separate PR for it, but thought I'd just squeeze it in here alongside its tests.

File: packages/react-form/src/hooks/test/form.test.tsx (hunk 8 above)
Comment (TheMallen) [reply packages/react-form/src/hooks/test/form.test.tsx]: confirmed locally that rebasing makes the skip not necessary.

File: packages/react-import-remote/src/test/ImportRemote.test.tsx (hunk 14 above)
Comment (cartogram) [reply packages/react-import-remote/src/test/ImportRemote.test.tsx]: No, but without the above async/awaits we get the act console warnings.

File: packages/react-form/src/hooks/test/form.test.tsx (hunk 8 above)
Comment (cartogram) [reply packages/react-form/src/hooks/test/form.test.tsx]: @TheMallen confirmed, removing the `skip` 👍

File: packages/react-form/src/hooks/test/form.test.tsx [hunk 21]
```diff
@@ -156,18 +160,18 @@ describe('useForm', () => {
         .find(TextField, {label: 'title'})!
         .trigger('onChange', 'tortoritos, the chip for turtles!');
 
-      wrapper
-        .find('button', {type: 'submit'})!
-        .trigger('onClick', clickEvent());
+      wrapper.act(() => {
+        wrapper
```
Comment (lemonmade) [inline packages/react-form/src/hooks/test/form.test.tsx]: I don't think you should need to wrap this in an act?

File: packages/react-form/src/hooks/test/form.test.tsx [hunk 22]
```diff
@@ -200,51 +205,58 @@ describe('useForm', () => {
 
       wrapper.find(TextField, {label: 'title'})!.trigger('onChange', '');
 
-      wrapper
-        .find('button', {type: 'submit'})!
-        .trigger('onClick', clickEvent());
+      wrapper.act(() => {
+        wrapper
```
Comment (lemonmade) [inline packages/react-form/src/hooks/test/form.test.tsx]: Same as above

File: packages/react-form/src/hooks/test/form.test.tsx [hunk 23]
```diff
@@ -200,51 +205,58 @@ describe('useForm', () => {
...
       const wrapper = mount(
-        <ProductForm data={fakeProduct()} onSubmit={submitSpy} />,
+        <ProductForm data={fakeProduct()} onSubmit={() => promise} />,
       );
 
-      await wrapper
-        .find('button', {type: 'submit'})!
-        .trigger('onClick', clickEvent());
+      await wrapper.act(async () => {
+        await wrapper
```
Comment (lemonmade) [inline packages/react-form/src/hooks/test/form.test.tsx]: Seems weird that sometimes you are waiting on calling this, sometimes you are waiting on the promise. Stick to one.

File: packages/react-graphql/src/hooks/tests/mutation.test.tsx [hunk 24]
```diff
@@ -62,8 +57,13 @@ describe.skip('useMutation', () => {
       {graphQL},
     );
 
-    mockMutation.find('button')!.trigger('onClick', undefined as any);
-    await graphQL.resolveAll();
+    mockMutation.act(() => {
+      mockMutation.find('button')!.trigger('onClick', undefined as any);
+    });
+
+    await mockMutation.act(async () => {
```
Comment (lemonmade) [inline packages/react-graphql/src/hooks/tests/mutation.test.tsx]: Is there harm in merging these act blocks?

File: packages/react-import-remote/src/test/ImportRemote.test.tsx (hunk 14 above)
Comment (lemonmade) [reply packages/react-import-remote/src/test/ImportRemote.test.tsx]: We should be able to await on the result of calling `load` instead then?

File: packages/react-form/src/hooks/test/form.test.tsx (hunk 21 above)
Comment (cartogram) [reply packages/react-form/src/hooks/test/form.test.tsx]: Removing fails the test actually. It can't find the element.

File: packages/react-form/src/hooks/test/form.test.tsx (hunk 22 above)
Comment (cartogram) [reply packages/react-form/src/hooks/test/form.test.tsx]: Same, same as above.
